gherkin8/src/main/java/io/cucumber/core/gherkin5/Gherkin8CucumberPickle.java

```diff
@@ -0,0 +1,100 @@
+package io.cucumber.core.gherkin5;
+
+import io.cucumber.core.gherkin.CucumberPickle;
+import io.cucumber.core.gherkin.CucumberStep;
+import io.cucumber.core.gherkin.StepType;
+import io.cucumber.gherkin.GherkinDialect;
+import io.cucumber.messages.Messages;
+import io.cucumber.messages.Messages.Location;
+import io.cucumber.messages.Messages.Pickle.PickleTag;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Wraps {@link Pickle} to avoid exposing the gherkin library to all of
+ * Cucumber.
+ */
+public final class Gherkin8CucumberPickle implements CucumberPickle {
+
+    private final Messages.Pickle pickle;
+    private final List<CucumberStep> steps;
+    private final URI uri;
+
+    Gherkin8CucumberPickle(Messages.Pickle pickle, URI uri, Messages.GherkinDocument document, GherkinDialect dialect) {
+        this.pickle = pickle;
+        this.uri = uri;
+        this.steps = createCucumberSteps(pickle, document, dialect);
+    }
+
+    private static List<CucumberStep> createCucumberSteps(Messages.Pickle pickle, Messages.GherkinDocument document, GherkinDialect dialect) {
+        List<CucumberStep> list = new ArrayList<>();
+        String previousGivenWhenThen = dialect.getGivenKeywords()
+            .stream()
+            .filter(s -> !StepType.isAstrix(s))
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException("No Given keyword for dialect: " + dialect.getName()));
+
+        for (Messages.Pickle.PickleStep step : pickle.getStepsList()) {
+            CucumberStep cucumberStep = new Gherkin8CucumberStep(step, document, dialect, previousGivenWhenThen);
+            if (cucumberStep.getStepType().isGivenWhenThen()) {
+                previousGivenWhenThen = cucumberStep.getKeyWord();
+            }
+            list.add(cucumberStep);
+        }
+        return list;
+    }
+
+    @Override
+    public String getLanguage() {
+        return pickle.getLanguage();
+    }
+
+    @Override
+    public String getName() {
+        return pickle.getName();
+    }
+
+    /**
+     * Returns the line in feature file of the Scenario this pickle was created
+     * from. If this pickle was created from a Scenario Outline this line is the
+     * line in the Example section used to fill in the place holders.
+     *
+     * @return line in the feature file
+     */
+    @Override
+    public int getLine() {
+        return pickle.getLocationsList().get(0).getLine();
+    }
+
+    /**
+     * Returns the line in feature file of the Scenario this pickle was created
+     * from. If this pickle was created from a Scenario Outline this line is the
+     *
+     * @return line in the feature file
+     */
+    @Override
+    public int getScenarioLine() {
+        List<Location> stepLocations = pickle.getLocationsList();
+        return stepLocations.get(stepLocations.size() - 1).getLine();
```

Comment:
Is this still correct? I had to swap this in `Gherkin8CucumberStep`

Reply:
Then I would appreciate a descriptive method to access the last line number. Perhaps we can call it `pickleLine` because it is the line that defines the pickle. It would remove one obstacle that keeps me from using plain pickles in Cucumber.

Reply:
I like that idea. Just like we've talked about adding a `GherkinDocumentsModel` class to cucumber-messages, we could add a `PickleModel` class to cucumber-messages as well.

In short - some utility classes for getting interesting data out of those dumb protobuf structs.

Reply:
Yes. Though I think utility classes are an anti-pattern when you own/can design the domain objects.

When ever you run into a pattern like:
```
Object dumb = ...
String propertyA = dumb.propertyA();
String propertyB = SmartUtils.propertyB(dumb)
```
It's better to rewrite it to a facade and use the facade everywhere. E.g:

```
Object dumb = ....
Smart  smart = new Smart(dumb);
String propertyA = smart.propertyA()
String propertyB = smart.propertyB();
```

Reply:
I think we’re after the same thing! This is basically the facade pattern.

Reply:
We could provide different flavours of facades around protobuf messages. One could process a stream of historical messages (stored in an event store like database, heck could even be Kafka). This way we could generate more interesting data like historical trends for execution time, flakiness, failure rates or even lead times from new to passing. The computed data could be stored in the database (if the computation is time consuming. This could be implemented as streams and pipes, allowing for composition.